x/incentive/keeper/reward_tracker.go

```diff
@@ -0,0 +1,295 @@
+// to adding rewards. In the sense that a FP would first receive active delegations sats
+// be properly initialized (creates current and historical reward structures in the store)
+// than will start to receive rewards for contributing.
+func (k Keeper) AddFinalityProviderRewardsForBtcDelegations(ctx context.Context, fp sdk.AccAddress, rwd sdk.Coins) error {
+	fpCurrentRwd, err := k.GetFinalityProviderCurrentRewards(ctx, fp)
+	if err != nil {
+		return err
+	}
+
+	fpCurrentRwd.AddRewards(rwd)
+	return k.setFinalityProviderCurrentRewards(ctx, fp, fpCurrentRwd)
+}
+
+// BtcDelegationActivated adds new amount of active satoshi to the delegation
+// and finality provider. Since it modifies the amount staked, it triggers
+// the creation of a new period, which initializes the FP, creates
+// historical reward tracker, withdraw the BTC delegation rewards to gauge
+// and initializes a new delegation with the just ended period.
+func (k Keeper) BtcDelegationActivated(ctx context.Context, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	return k.btcDelegationModifiedWithPreInitDel(ctx, fp, del, func(ctx context.Context, fp, del sdk.AccAddress) error {
+		return k.addDelegationSat(ctx, fp, del, amtSat)
+	})
+}
+
+// BtcDelegationUnbonded it modifies the total amount of satoshi staked
+// for the delegation (fp, del) and for the finality provider by subtracting.
+// Since it modifies the active amount it triggers the increment of fp period,
+// creationg of new historical reward, withdraw of rewards to gauge
+// and initialization of a new delegation.
+// It errors out if the unbond amount is higher than the total amount staked.
+func (k Keeper) BtcDelegationUnbonded(ctx context.Context, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	return k.btcDelegationModifiedWithPreInitDel(ctx, fp, del, func(ctx context.Context, fp, del sdk.AccAddress) error {
+		return k.subDelegationSat(ctx, fp, del, amtSat)
+	})
+}
+
+// FpSlashed a slashed finality provider should withdraw all the rewards
+// available to it, by iterating over all the delegations for this FP
+// and sending to the gauge. After the rewards are removed, it should
+// delete every rewards tracker value in the store related to this slashed
+// finality provider.
+func (k Keeper) FpSlashed(ctx context.Context, fp sdk.AccAddress) error {
+	// finalize the period to get a new history with the current rewards available
+	endedPeriod, err := k.IncrementFinalityProviderPeriod(ctx, fp)
+	if err != nil {
+		return err
+	}
+
+	// remove all the rewards available from the just ended period
+	keysBtcDelRwdTracker := make([][]byte, 0)
+	if err := k.IterateBTCDelegationRewardsTracker(ctx, fp, func(fp, del sdk.AccAddress) error {
+		keysBtcDelRwdTracker = append(keysBtcDelRwdTracker, del.Bytes())
+		return k.CalculateBTCDelegationRewardsAndSendToGauge(ctx, fp, del, endedPeriod)
+	}); err != nil {
+		return err
+	}
```

Comment:
Yeah tbh I'm not sure if this is necessary, and this iterating process might be some perf hotspot (It iterates over all BTC delegations under a slashed FP).